kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/testplugins/AuditLoggerFilter.java

```diff
@@ -0,0 +1,203 @@
+            // TODO netty has a ChannelId which we should probably use??
+            object.put("clientCertSig", context.clientTlsContext().flatMap(ClientTlsContext::clientCertificate)
+                    .map(X509Certificate::getSignature).orElse(null));
+            object.put("clientId", header.clientId());
+            object.put("correlationId", header.correlationId());
+            object.put("apiKey", apiKey.name);
+            if (logRequest) {
+                // TODO don't do this for SaslAuthenticate, to avoid logging private credentials
+                object.set("request", requestJson(apiKey, header.requestApiVersion(), request, false));
+            }
+            log(object);
+        }
+        if (loggableResponses.get(apiKey)) {
+            requestVersions.put(header.correlationId(), header.requestApiVersion());
+        }
+
+        return context.forwardRequest(header, request);
+    }
+
+    @Override
+    public CompletionStage<ResponseFilterResult> onResponse(ApiKeys apiKey,
+                                                            ResponseHeaderData header,
+                                                            ApiMessage response,
+                                                            FilterContext context) {
+        Boolean logResponse = loggableResponses.get(apiKey);
+        if (logResponse != null) {
+            var object = JsonNodeFactory.instance.objectNode();
+            long nanosSinceInit = System.nanoTime() - initNanos;
+            object.put("nanosSinceInit", nanosSinceInit);
+            object.put("currentTimeMillis", currentTimeMillis(nanosSinceInit));
+            // object.put("channelDescriptor", context.channelDescriptor());
+            object.put("clientCertSig", context.clientTlsContext().flatMap(ClientTlsContext::clientCertificate)
+                    .map(X509Certificate::getSignature).orElse(null));
+            // TODO object.put("serverPrincipal", serverPrincipal.getName());
+            object.put("correlationId", header.correlationId());
+            object.put("apiKey", apiKey.name);
+            if (logResponse) {
+                short requestApiVersion = requestVersions.remove(header.correlationId());
+                // TODO don't do this for SaslAuthenticate, to avoid logging private credentials
+                object.set("response", responseJson(apiKey, requestApiVersion, response, false));
+            }
+            log(object);
+        }
+        return context.forwardResponse(header, response);
+    }
+
+    @Nullable
+    private static JsonNode requestJson(ApiKeys apiKey, short requestApiVersion, ApiMessage request, boolean serializeRecords) {
+        return switch (apiKey) {
+            // TODO generate a method to do this using the krpc plugin
+            case API_VERSIONS -> ApiVersionsRequestDataJsonConverter.write((ApiVersionsRequestData) request, requestApiVersion, serializeRecords);
```

Comment:
Does io.kroxylicious.test.requestresponsetestdef.KafkaApiMessageConverter fit the bill?